File: packages/cppproxy/examples/person.pl
:- use_module(parms).

:- use_module(library(typedef)).
:- use_module(library(cpp_interface)).
:- use_module(library(cpp_codegen)).
:- use_module(library(cpp_server)).

		 /*******************************
		 *	       TYPES		*
		 *******************************/

:- type first_name = atom.
:- type last_name  = atom.
:- type age        = integer.

:- type person -> person(first_name, last_name, age).

:- cpp_callable
	add_person(+person),
	find_person_younger_than(+age, -person) = [zero_or_more].

add_person(person(FirstName, LastName, Age)) :-
	assert(person(FirstName, LastName, Age)).

find_person_younger_than(MaxAge, person(FirstName, LastName, Age)) :-
	person(FirstName, LastName, Age),
	Age =< MaxAge.

		 /*******************************
		 *	      GENERATE		*
		 *******************************/

make_client :-
	format('Creating wrapper ...'), flush_output,
	cpp_server_code('person_proxy.h',
			[ server_class('PersonProxy')
			]),
	format('ok.~ncompiling C++ code ...'), flush_output,
	compile('person.cpp'),
	format('ok.~n').


		 /*******************************
		 *	       SERVER		*
		 *******************************/

start_server :-
	port(Port),
	cpp_server([port(Port)]).


		 /*******************************
		 *	       DATA		*
		 *******************************/

:- dynamic
	person/3.

person('John',  'Jonsson', 34).
person('Mary',  'Smith',   28).
person('Bob',   'Smith',   58).
person('Terry', 'Waits',   52).
person('Tim',   'Seldon',  14).

File: packages/cppproxy/examples/person.cpp
#include "../SWI-proxy.cpp"

#ifndef PORT
#define PORT 4224
#endif

class person
{ 
public:
  char *first_name;
  char *last_name;
  int age;

  person()
  { first_name = NULL;
    last_name = NULL;
    age = -1;
  };
  ~person()
  { if ( first_name ) free(first_name);
    if ( last_name ) free(last_name);
  }

  char *get_first_name() const { return first_name; }
  char *get_last_name() const  { return last_name; }
  long  get_age() const  { return age; }

  void initialize(string fn, string ln, long years)
  { if ( first_name ) free(first_name);
    if ( last_name  ) free(last_name);

    first_name = strdup(fn.c_str());
    last_name  = strdup(ln.c_str());
    age = years;
  }
};

#include "person_proxy.h"

int
main(int argc, char **argv)
{ PersonProxy proxy("localhost", PORT);
  long maxage;
  
  if ( argc == 2 )
  { maxage = atoi(argv[1]);
  } else
  { maxage = 50;
  }

  try
  { person p;
    p.first_name = "Jan";
    p.last_name = "Wielemaker";
    p.age = 45;

    proxy.add_person(p);
  } catch ( PlException &ex )
  { cerr << ex << endl;
    return 1;
  }

  try
  { find_person_younger_than q(&proxy);
    person p;

    while( q.next_solution(maxage, p) )
      cout << p.first_name << " " << p.last_name << " " << p.age << endl;
  } catch ( PlException &ex )
  { cerr << ex << endl;
    return 1;
  }

  return 0;
}
